### PR TITLE
[DO_NOT_MERGE][NP-5016] Provide "Save as PDF" button for business onboarding wizard to save the PDF to fileDAO

### PR DESCRIPTION
**Pull request by thepinkcacao**
*Foam PR: https://github.com/kgrgreer/foam3/pull/780
*Nanopay PR : https://github.com/nanoPayinc/NANOPAY/pull/13219
*Ticket: https://nanopay.atlassian.net/browse/NP-5016

*Description: 
-  Added a button called “saveToPdf” in business onboarding wizard. 

> convert HTML page to PDF file 

> save the file to fileDAO

-  Added a permission to delete a file (mark a deleted file) to keep only the recent generated file in fileDAO. 

> only Fraud-ops can delete a file  

*Screenshots and pdf file example:
 1) PDF file example(Business Onboarding)
[businessOnboardingInfo.pdf](https://github.com/kgrgreer/foam3/files/6977494/businessOnboardingInfo.pdf)

 2) Permission to fraud - ops ( can delete a file)
can delete (mark) file with fraudops

 3) No permission to user (cannot delete a file)
cannot delete(mark) file with a user

### CODE REVIEW

File: src/foam/nanos/fs/File.js [hunk 1]
```diff
@@ -17,7 +17,8 @@ foam.CLASS({
 
   implements: [
     'foam.nanos.auth.Authorizable',
-    'foam.nanos.auth.ServiceProviderAware'
+    'foam.nanos.auth.ServiceProviderAware',
+    'foam.nanos.auth.LifecycleAware'
```
Comment (KolombetM) [inline src/foam/nanos/fs/File.js]: This should be in alphabetical order

Comment (thepinkcacao) [reply src/foam/nanos/fs/File.js]: right, I will change this. Thank you!

**Review by Mykola97 — APPROVED** [review]

File: src/foam/nanos/fs/File.js [hunk 2]
```diff
@@ -267,7 +279,8 @@ foam.CLASS({
       name: 'authorizeOnDelete',
       javaCode: `
         AuthService auth = (AuthService) x.get("auth");
-        if ( ! auth.check(x, "file.remove." + getId()) ) {
+
+        if ( ! auth.check(x, "file.remove.*") ) {
```
Comment (lchanmann) [inline src/foam/nanos/fs/File.js]: @Mykola97 @thepinkcacao  why not "file.remove.id"?

**Review by lchanmann — CHANGES_REQUESTED** [review]

File: src/foam/nanos/fs/File.js (hunk 2 above)
Comment (thepinkcacao) [reply src/foam/nanos/fs/File.js]: @lchanmann  does id mean "fileID" or "UserID"?

File: src/foam/nanos/fs/File.js [hunk 3]
```diff
@@ -267,7 +279,8 @@ foam.CLASS({
       name: 'authorizeOnDelete',
       javaCode: `
         AuthService auth = (AuthService) x.get("auth");
-        if ( ! auth.check(x, "file.remove." + getId()) ) {
+
+        if ( ! auth.check(x, "file.remove.id") ) {
```
Comment (lchanmann) [inline src/foam/nanos/fs/File.js]: should be `"file.remove." + getId()`, will also want to allow users to remove the files they upload and own.

Comment (thepinkcacao) [reply src/foam/nanos/fs/File.js]: okay, thank you.

File: src/foam/nanos/fs/File.js [hunk 4]
```diff
@@ -267,7 +283,15 @@ foam.CLASS({
...
       javaCode: `
         AuthService auth = (AuthService) x.get("auth");
-        if ( ! auth.check(x, "file.remove." + getId()) ) {
+        Boolean hasPermission = auth.check(x, "file.remove." + getId());
+
+        // find a file owner
+        // creating a file uses businessID for the file owner
+        DAO businessDAO = (DAO) x.get("businessDAO");
+        Business fileOwner = (Business) businessDAO
+          .find(EQ(Business.ID, getOwner()));
```
Comment (lchanmann) [inline src/foam/nanos/fs/File.js]: should check if the current user is the owner and the real user is the createdByAgent. See an example at https://github.com/kgrgreer/foam3/blob/d5039484d241b9ee6643d432c9c1b39385e26b80/src/foam/nanos/ticket/Ticket.js#L351-L356.

File: src/foam/nanos/fs/File.js [hunk 5]
```diff
@@ -267,7 +282,11 @@ foam.CLASS({
       name: 'authorizeOnDelete',
       javaCode: `
         AuthService auth = (AuthService) x.get("auth");
-        if ( ! auth.check(x, "file.remove." + getId()) ) {
+        Boolean hasPermission = auth.check(x, "file.remove." + this.getId());
+        Subject subject = (Subject) x.get("subject");
+        User user = subject.getRealUser();
+
+        if ( ! hasPermission && user.getId() != this.getCreatedBy() ) {
```
Comment (lchanmann) [inline src/foam/nanos/fs/File.js]: Please use subject.getUser() instead then check it against the file owner.
```suggestion
        User user = subject.getUser();

        if ( ! hasPermission && user.getId() != this.getOwner() ) {
```

Comment (thepinkcacao) [reply src/foam/nanos/fs/File.js]: okay, thank you.

**Pull request closed without merging**